### ISSUE TITLE
Implement `@Contravariant` and improve `@Covariant` syntax. Enforce correct usage.

### ISSUE THREAD

**Issue by justmin** [typetools/checker-framework#1955]
Both functions in below example produce CF compile-time error and this is correct.
But it would be nice if we could tell CF some assumptions about generic args possible usages: as return values only or as parameters only.
In such case this code could pass compilation and it will be correct.
```
public class InheritanceTest {
    public interface Getter<R>{
        R get();
    }

    public interface Consumer<P>{
        void accept(P arg);
    }

    Getter<@Nullable String> foo1(Getter<String> stringGetter){
        return stringGetter;
    }

    Consumer<String> foo2(Consumer<@Nullable String> consumer){
        return consumer;
    }
}

```

**Comment by wmdietl** [comment]
The Checker Framework uses the same mechanism as Java: wildcards a.k.a. use-site variance annotations.
The following works:

```java
import org.checkerframework.checker.nullness.qual.Nullable;

public class InheritanceTest {
    public interface Getter<R> {
        R get();
    }

    public interface Consumer<P> {
        void accept(P arg);
    }

    Getter<? extends @Nullable String> foo1(Getter<String> stringGetter) {
        return stringGetter;
    }

    Consumer<? super String> foo2(Consumer<@Nullable String> consumer) {
        return consumer;
    }
}
```

Yes, it is a bit awkward to use an `extends String`, given that String is a final class.

We also support a declaration-site variance annotation: [Covariant](https://github.com/typetools/checker-framework/blob/master/framework/src/main/java/org/checkerframework/framework/qual/Covariant.java)

That makes the following possible:

```java
import org.checkerframework.checker.nullness.qual.Nullable;
import org.checkerframework.framework.qual.Covariant;


public class InheritanceTest {
    @Covariant(0)
    public interface Getter<R> {
        R get();
    }

    public interface Consumer<P> {
        void accept(P arg);
    }

    Getter<@Nullable String> foo1(Getter<String> stringGetter) {
        return stringGetter;
    }

    Consumer<? super String> foo2(Consumer<@Nullable String> consumer) {
        return consumer;
    }
}
```

Is this the kind of mechanism you had in mind?
We currently do not enforce correct usage of Covariant type parameters, i.e. that they are only used in return positions.
We also don't yet have a `@Contravariant` annotation, but it should be relatively straightforward to add support for that.

**Comment by justmin** [comment]
Covariant works for me, thanks!
But IMHO it look syntactically ugly - I prefer another style of annotations:

```
public interface Getter<@Covariant R> {
        R get();
}

```

**Comment by mernst** [comment]
This issue mixes together three different enhancements.
I'm closing it in favor of #1969, #1970, and #1971.
